x-pack/plugins/dataset_quality/public/state_machines/dataset_quality_controller/src/types.ts

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { DoneInvokeEvent } from 'xstate';
+import { DataStreamDetails } from '../../../../common/data_streams_stats/data_stream_details';
+import { Direction, SortField } from '../../../hooks';
+import { DegradedDocsStat } from '../../../../common/data_streams_stats/malformed_docs_stat';
+import {
+  DataStreamDegradedDocsStatServiceResponse,
+  DataStreamStatServiceResponse,
+} from '../../../../common/data_streams_stats';
+import { DataStreamStat } from '../../../../common/data_streams_stats/data_stream_stat';
+
+export interface FlyoutDataset {
+  rawName: string;
+  type: string;
+  name: string;
+  namespace: string;
+  title: string;
+  integration?: {
+    name: string;
+    title: string;
+    version: string;
+  };
+}
+
+interface TableCriteria {
+  page: number;
+  rowsPerPage: number;
+  sort: {
+    field: SortField;
+    direction: Direction;
+  };
+}
+
+export interface WithTableOptions {
+  table: TableCriteria;
+}
+
+export interface WithFlyoutOptions {
+  flyout: {
+    dataset?: FlyoutDataset;
+    datasetDetails?: DataStreamDetails;
+  };
+}
+
+export interface WithDataStreamStats {
+  dataStreamStats: DataStreamStat[];
+}
+
+export interface WithDegradedDocs {
+  degradedDocStats: DegradedDocsStat[];
+}
+
+export interface WithDatasets {
+  datasets?: string;
+}
+
+export type DefaultDatasetQualityControllerState = WithTableOptions &
+  WithDataStreamStats &
+  WithDegradedDocs &
+  WithFlyoutOptions &
+  WithDatasets;
+
+type DefaultDatasetQualityStateContext = DefaultDatasetQualityControllerState &
+  Partial<WithFlyoutOptions>;
+
+export type DatasetQualityControllerTypeState =
+  | {
+      value: 'datasets.fetching';
+      context: DefaultDatasetQualityStateContext;
+    }
+  | {
+      value: 'datasets.loaded';
+      context: DefaultDatasetQualityStateContext;
+    }
+  | {
+      value: 'datasets.loaded.idle';
+      context: DefaultDatasetQualityStateContext;
+    }
+  | {
+      value: 'datasets.loaded.flyoutOpen.fetching';
+      context: DefaultDatasetQualityStateContext;
+    }
+  | {
+      value: 'datasets.loaded.flyoutOpen';
+      context: DefaultDatasetQualityStateContext;
+    }
+  | {
+      value: 'degradedDocs.fetching';
+      context: DefaultDatasetQualityStateContext;
+    };
+
+export type DatasetQualityControllerContext = DatasetQualityControllerTypeState['context'];
+
+export type DatasetQualityControllerEvent =
+  | {
+      type: 'UPDATE_TABLE_CRITERIA';
+      criteria: TableCriteria;
+    }
+  | {
```

Comment:
**nit**: let's keep events as atomic discriminated unions to keep the transitions granular and easily update any event payload.

Reply:
what do you mean?

Reply:
```ts
| {
    type: 'OPEN_FLYOUT';
    dataset: FlyoutDataset;
  }
| {
    type: 'SELECT_NEW_DATASET';
    dataset: FlyoutDataset;
  }

```

Reply:
Addressed in [66b7dbb](https://github.com/elastic/kibana/pull/174906/commits/66b7dbbeebe835ec361d8b7896ceb664c514e304).